File: sdk/python/clampd/contract_hash.py
"""Canonical tool-contract hash — wire-identical to the TypeScript SDK.

Same formula as ``sdk/typescript/src/contract-hash.ts`` so a tool
registered from Python and invoked from TS (or vice versa) produces the
same ``descriptor_hash`` — rug-pull detection at the gateway can then
distinguish "admin approved this exact contract" from "someone changed
the contract".

Formula:
    sha256(json.dumps(
        {"name": ..., "description": ..., "parameters": ...},
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ))

``sort_keys=True`` + ``separators=(",", ":")`` matches the TS SDK's
``sortedStringify`` output byte-for-byte, modulo unicode escaping
(``ensure_ascii=False`` keeps raw characters like the TS ``JSON.stringify``
default). Both ends emit the same canonicalisation.
"""
from __future__ import annotations

import hashlib
import json
from typing import Any


def contract_hash(name: str, description: str, parameters: Any) -> str:
    """Compute the canonical descriptor hash for a tool contract.

    Args:
        name: Tool name as it appears in proxy requests (e.g. ``"db.query"``).
        description: Human-readable description — participates in the hash,
            so a change here fires a rug-pull alert. Empty string is fine
            when the caller has no description to offer.
        parameters: JSON-Schema object describing the tool's parameters.
            Use ``{}`` when the caller has no schema. Must be
            JSON-serialisable.

    Returns:
        64-character lowercase hex string (SHA-256).
    """
    canonical = json.dumps(
        {
            "name": name,
            "description": description or "",
            "parameters": parameters if parameters is not None else {},
        },
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    )
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()

File: sdk/typescript/src/contract-hash.ts
/**
 * Canonical contract hash for a tool descriptor.
 *
 * Single source of truth: the hash is a SHA-256 over the
 * JSON-canonicalised tuple `{name, description, parameters}`. Must be
 * computed identically by:
 *
 *   - The SDK's `guard()` (when wrapping a tool call)
 *   - The SDK's `registerTool()` (when declaring a tool up-front)
 *   - dashboard-api's `/tool-descriptors/:org/register` endpoint
 *   - Any future admin tooling that writes to `tool_descriptors.descriptor_hash`
 *
 * Prior to SDK v0.13.0 the SDK used `sha256(toolName + ":" + fn.toString())`
 * and the server used `sha256(org_id + ":" + name + ":sdk")`. Neither
 * agreed with the other, so runtime rug-pull detection could not
 * distinguish "admin approved a different hash" from "caller produced a
 * different hash because `fn.toString()` happens to be different today".
 * Both routes now pass through this function so the hash is content-
 * addressed: anything that changes the external tool contract (name,
 * description, parameter schema) changes the hash; nothing else does.
 */
import { createHash } from "node:crypto";
import { sortedStringify } from "./config.js";

export interface ToolContract {
  /** Canonical tool name (e.g. "database.query"). */
  name: string;
  /** Human-readable description (shown to the LLM and in the dashboard). */
  description: string;
  /** JSON-Schema describing the tool's parameters. May be `{}` when unknown. */
  parameters: object;
}

/**
 * Compute the canonical descriptor hash for a tool contract.
 *
 * Deterministic: identical input always yields the same output. The
 * canonicalisation is key-sorted JSON with no whitespace — reordering
 * an object's fields does not change the hash.
 */
export function contractHash(contract: ToolContract): string {
  const canonical = sortedStringify({
    name: contract.name,
    description: contract.description ?? "",
    parameters: contract.parameters ?? {},
  });
  return createHash("sha256").update(canonical).digest("hex");
}

/**
 * Compute the per-call binding hash for a scope token.
 *
 * Wire-identical to `ag_common::contract_hash::call_binding` (Rust) and the
 * Python SDK's `call_binding`. Binds a scope token to the EXACT call it
 * authorized — the tool name plus the call params — so a token granted for one
 * call cannot be replayed for a different one. The tool side recomputes this
 * from the call it received and rejects the token on mismatch.
 *
 * `binding = sha256( canonical_json({ params, tool }) )` — key-sorted JSON,
 * no whitespace, raw UTF-8. `tool` is the name exactly as sent in the proxy
 * request (not re-canonicalised), so both ends agree without shared rules.
 */
export function callBinding(tool: string, params: unknown): string {
  const canonical = sortedStringify({ params: params ?? {}, tool });
  return createHash("sha256").update(canonical).digest("hex");
}
